File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.java
package com.novoda.storagepathfinder.demo;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.novoda.storagepathfinder.DeviceStorageInspector;
import com.novoda.storagepathfinder.StoragePath;

import java.util.ArrayList;
import java.util.List;

public class LandingActivity extends AppCompatActivity {

    private static final String ASSET_NAME = "10MB.zip";

    private AssetCloner assetCloner;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_landing);

        assetCloner = DemoDependenciesFactory.createAssetCloner(getApplicationContext());

        DeviceStorageInspector storageInspector = DeviceStorageInspector.builder(this).build();

        List<StoragePath> deviceStoragePaths = new ArrayList<>();

        deviceStoragePaths.addAll(storageInspector.getPrimaryStorageBasePaths());
        deviceStoragePaths.addAll(storageInspector.getPrimaryStorageApplicationPaths());
        deviceStoragePaths.addAll(storageInspector.getSecondaryStorageBasePaths());
        deviceStoragePaths.addAll(storageInspector.getSecondaryStorageApplicationPaths());

        RecyclerView recyclerView = findViewById(R.id.device_storage_roots);
        recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        StoragePathsAdapter storagePathsAdapter = new StoragePathsAdapter(
                getLayoutInflater(),
                deviceStoragePaths,
                onAddFileClicked
        );
        recyclerView.setAdapter(storagePathsAdapter);
    }

    private final StoragePathViewHolder.Listener onAddFileClicked = deviceStoragePath -> {
        String pathToCloneTo = deviceStoragePath.getPathAsString() + "/" + ASSET_NAME;
        assetCloner.cloneAsset(ASSET_NAME, pathToCloneTo);
    };

}

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt
package com.novoda.storagepathfinder.demo

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import com.novoda.storagepathfinder.StoragePath
import kotlinx.android.synthetic.main.activity_landing.*
import java.util.*

class LandingActivity : AppCompatActivity() {

    private var assetCloner: AssetCloner? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_landing)

        assetCloner = DemoDependenciesFactory.createAssetCloner(applicationContext)
        val storageInspector = DemoDependenciesFactory.createStorageInspector(applicationContext)

        val deviceStoragePaths = ArrayList<StoragePath>()
        deviceStoragePaths.add(storageInspector.primaryStorageBasePath)
        deviceStoragePaths.add(storageInspector.primaryStorageApplicationPath)
        deviceStoragePaths.addAll(storageInspector.secondaryStorageBasePath)
        deviceStoragePaths.addAll(storageInspector.secondaryStorageApplicationPath)

        device_storage_roots.layoutManager = LinearLayoutManager(applicationContext)
        device_storage_roots.adapter = StoragePathsAdapter(
                layoutInflater,
                deviceStoragePaths,
                onAddFileClicked
        )
    }

    private val onAddFileClicked: (StoragePath) -> Unit = { deviceStoragePath ->
        val pathToCloneTo = deviceStoragePath.pathAsString + "/" + ASSET_NAME
        assetCloner?.cloneAsset(ASSET_NAME, pathToCloneTo)
    }

    companion object {
        private const val ASSET_NAME = "10MB.zip"
    }

}
